instaloader/instaloader.py

```diff
@@ -223,20 +233,15 @@ def __init__(self,
                  compress_json: bool = True,
                  post_metadata_txt_pattern: Optional[str] = None,
                  storyitem_metadata_txt_pattern: Optional[str] = None,
-                 max_connection_attempts: int = 3,
-                 request_timeout: float = 300.0,
-                 rate_controller: Optional[Callable[[InstaloaderContext], RateController]] = None,
                  resume_prefix: Optional[str] = "iterator",
                  check_resume_bbd: bool = True,
                  slide: Optional[str] = None,
-                 fatal_status_codes: Optional[List[int]] = None,
-                 iphone_support: bool = True,
                  title_pattern: Optional[str] = None,
-                 sanitize_paths: bool = False):
+                 sanitize_paths: bool = False,
+                 context: Optional[InstaloaderContext] = None
+                 ):
```

Comment:
We cannot remove parameters from the Instaloader constructor, this would potentially break existing code that uses Instaloader.

Reply:
If `context` is passed in, we can simply ignore any removed parts of `__init__`.
But since there's a factory there anyway, just use them in there. This will add backward compatibility.

```python
def __init__(self,
             sleep: bool = True,
             quiet: bool = False,
             user_agent: Optional[str] = None,
             dirname_pattern: Optional[str] = None,
             filename_pattern: Optional[str] = None,
             download_pictures=True,
             download_videos: bool = True,
             download_video_thumbnails: bool = True,
             download_geotags: bool = False,
             download_comments: bool = False,
             save_metadata: bool = True,
             compress_json: bool = True,
             post_metadata_txt_pattern: Optional[str] = None,
             storyitem_metadata_txt_pattern: Optional[str] = None,
             max_connection_attempts: int = 3,
             request_timeout: float = 300.0,
             rate_controller: Optional[Callable[[InstaloaderContext], RateController]] = None,
             resume_prefix: Optional[str] = "iterator",
             check_resume_bbd: bool = True,
             slide: Optional[str] = None,
             fatal_status_codes: Optional[List[int]] = None,
             iphone_support: bool = True,
             title_pattern: Optional[str] = None,
             sanitize_paths: bool = False,
             context: Optional[InstaloaderContext] = None
             ):
        self.context = context or instaloader_context_factory(not sleep, quiet, user_agent,
                 max_connection_attempts, request_timeout, rate_controller,
                 fatal_status_codes, iphone_support)
        ...
```

Reply:
@Henkhogan, I'm playing with your fork now. Firstly I've thought, that would be easier to pass `Callable` for `session_factory`. But context factory much greater, since it gives rights to rewrite used user-agent, cookies and logic of requests.

Reply:
My idea is to create a new base class that Instaloader inherits from. That will allow to not break anything while the new base class can be used by whoever needs more control of the context

Reply:
Another approach that would allow a clean extension of how the sessions are handled would be to refactor the requests.Session handling that we have in instaloadercontext.py into an own session class, extending requests.Session, and having a parameter for the session factory, similar as it is done with the rate controller (https://instaloader.github.io/module/instaloadercontext.html#instaloader.RateController). This provides an easily-overridable default.